### PR TITLE
New python binds

### PR DESCRIPTION
**Pull request by igo95862**
#701 

This is a big lump of changes made by copy pasting.

It should be possible to make commits separated if there is interest.

### CODE REVIEW

**Comment by igo95862** [comment]
Some questions unsolved:

* Integration with build system. I could not find any CMAKE files in the old python binds. Right now it uses the default `setuptools`.
* readthedocs: I tested it on [my repository](https://notcurses-python.readthedocs.io/en/latest/?) but I am not sure how it can be integrated on readthedocs in the scope of the large project. I saw you wanted to migrate to readthedocs. Maybe its possible to setup a project wide sphinx documentation that will import from the python documentation in one of the subsections.

**Comment by dankamongmen** [comment]
I will take a look at this ASAP, though it might not be until this weekend. Thanks so much @igo95862 !

As to your questions:
(1) we used to build the python out of the main cmake build, but i pulled it out a few months back
(2) i will take a look at it. i have a branch open right now where i've started using mdbook, on the advice of @joseluis

**Comment by dankamongmen** [comment]
Alright, reviewing this today! Sorry for the delay!

**Comment by dankamongmen** [comment]
I'm very much liking what I see so far. I can bring this up to speed with the current head, or you can. I think this will likely be merged largely unchanged.

**Comment by igo95862** [comment]
Can you rebase it? I've been busy with another project.

If you have any questions on how to extend the binds feel free to ask.

**Comment by dankamongmen** [comment]
no problem, will do

**Comment by dankamongmen** [comment]
I've brought it up to date in `igo95862-master`

File: python/README.md
Comment (dankamongmen) [inline python/README.md:1]: This doesn't seem to have been replaced by .readthedocs.yml or anything else. Updating the author is certainly called for, but ought this go away entirely? Isn't it used by pypi?

File: python/docs/nc_direct.rst
Comment (dankamongmen) [inline python/docs/nc_direct.rst:1]: Do these have to be kept up-to-date by hand, or are they automatically generated?

File: python/examples/001-init-notcurses.py
Comment (dankamongmen) [inline python/examples/001-init-notcurses.py:22]: Why the sleep here? Were you seeing failures otherwise? I'd rather not introduce a second of latency to every application that uses us. That would likely cause me to reach for a different library.

Comment (dankamongmen) [reply python/examples/001-init-notcurses.py:22]: ohhhhhhhhhhhhhh, this is example code. OK, what I would do, then, is use the `NOTCURSES_NO_ALTERNATE_SCREEN` ncoption. I assume you have this `sleep()` to make the output actually visible? Yeah, use the `NO_ALTERNATE_SCREEN`, please.

**Review by dankamongmen — CHANGES_REQUESTED** [review]

File: python/setup.py
Comment (dankamongmen) [inline python/setup.py:7]: We can't do this; Debian and Fedora both require a man page to be installed for every installed executable. I can add it back in for you, but we can't remove it. I'll change the authorship in them.

File: python/notcurses-pydemo
Comment (dankamongmen) [inline python/notcurses-pydemo:1]: Ahh, I see, the two demo scripts have both been removed. Hrmmm. I'm not crazy about that.

Comment (igo95862) [reply python/notcurses-pydemo:1]: I think I moved both of them in to the examples.

File: python/examples/001-init-notcurses.py
Comment (igo95862) [reply python/examples/001-init-notcurses.py:22]: Basically after you render something you need sleep for it to stay on screen. Otherwise once you reach the end of the script the context and all planes will get deferenced and the screen will revert to original state. It should be possible to include more advanced event loop in the future. The input test example has a state tied to inputs so it does not use sleep.

**Comment by dankamongmen** [comment]
OK, overall this looks like a tremendous advance over what I had, by someone who actually knows Python. I can't thank you enough for this great work.

Unfortunately, there are two hard blockers on this, and one soft blocker:
* [ ] I'd like the two demonstration scripts, or rough equivalents, to remain, especially as they're part of our meagre continuous integration. Without them, it looks like we have no testing available whatsoever.
* [ ] Debian and Fedora both require a man page for every executable, so we'll either need restore them (if the existing scripts are restored), or write new ones (for new demonstration scripts)
* [ ] I'd *much* rather see you use `NOTCURSES_NO_ALTERNATE_SCREEN` rather than `sleep(N)` in your example code, especially since it's example code, i.e. the code most people will see first

Also, I see the following warning when I try to run locally:

```
[schwarzgerat](1) $ python3 setup.py build
running build
running build_py
creating build
creating build/lib.linux-x86_64-3.8
creating build/lib.linux-x86_64-3.8/notcurses
copying notcurses/__init__.py -> build/lib.linux-x86_64-3.8/notcurses
copying notcurses/_notcurses.py -> build/lib.linux-x86_64-3.8/notcurses
copying notcurses/notcurses.py -> build/lib.linux-x86_64-3.8/notcurses
running build_ext
building 'notcurses/_notcurses' extension
creating build/temp.linux-x86_64-3.8
creating build/temp.linux-x86_64-3.8/notcurses
x86_64-linux-gnu-gcc -pthread -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O2 -Wall -g -fstack-protector-strong -Wformat -Werror=format-security -g -fwrapv -O2 -g -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -fPIC -I/usr/include/python3.8 -c notcurses/_notcurses.c -o build/temp.linux-x86_64-3.8/notcurses/_notcurses.o
notcurses/_notcurses.c: In function ‘_notcurses_context_get_input_blocking’:
notcurses/_notcurses.c:595:20: warning: comparison of integer expressions of different signedness: ‘char32_t’ {aka ‘unsigned int’} and ‘int’ [-Wsign-compare]
  595 |     if (code_point != -1)
      |                    ^~
x86_64-linux-gnu-gcc -pthread -shared -Wl,-O1 -Wl,-Bsymbolic-functions -Wl,-z,relro -g -fwrapv -O2 -Wl,-z,relro -g -fwrapv -O2 -g -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 build/temp.linux-x86_64-3.8/notcurses/_notcurses.o -o build/lib.linux-x86_64-3.8/notcurses/_notcurses.cpython-38-x86_64-linux-gnu.so
[schwarzgerat](0) $ 
```
Other than that, I think it goes in unmodified. Beautiful work.

File: python/setup.py
Comment (igo95862) [reply python/setup.py:7]: Feel free to modify. I only included the setup.py so it can be built quickly. Never expected it to stay the unchanged.

File: python/docs/nc_direct.rst
Comment (igo95862) [reply python/docs/nc_direct.rst:1]:
```
NcDirect documentation
======================

.. toctree::
    :maxdepth: 2
    :caption: Contents:

.. autoclass:: notcurses.NcDirect
    :members:
    :special-members: __init__
```

Automatic: new members of classes except the ones that start with double underscore.

New classes have to be added manually.

**Comment by igo95862** [comment]
> I'd like the two demonstration scripts, or rough equivalents, to remain, especially as they're part of our meagre continuous integration. Without them, it looks like we have no testing available whatsoever.

I put the demonstration script in to the example folder. The example 004 is the exact copy of the demonstration script. If you want to the examples can be moved in to the root folder of python binds.

> Debian and Fedora both require a man page for every executable, so we'll either need restore them (if the existing scripts are restored), or write new ones (for new demonstration scripts)

Are you sure you want to include them in to the package as binaries? In my opinion if you include them in package they should go to somewhere like `/usr/share/notcurses_py/examples`.

> I'd much rather see you use NOTCURSES_NO_ALTERNATE_SCREEN rather than sleep(N) in your example code, especially since it's example code, i.e. the code most people will see first

I had no idea `NOTCURSES_NO_ALTERNATE_SCREEN` was a thing.

**Comment by WSLUser** [comment]
Can you add a pyproject.toml for use with Poetry in the python root directpry? Dependency management is far better with Poetry than pip (even the new pipx) https://github.com/python-poetry/poetry

Also were these run through pylint, mypy and black to ensure standards compliance? 

I can add a linter Github Action using Github's Super Linter to help for reporting linter errors but likely configuration files will be desired to suppress certain warnings which maintainers of this project will need to add if desired. If there's interest in this, I'll work on a PR for it.

**Comment by dankamongmen** [comment]
> Can you add a pyproject.toml for use with Poetry in the python root directory? Dependency management is far better with Poetry than pip (even the new pipx) https://github.com/python-poetry/poetry
> 
> Also were these run through pylint, mypy and black to ensure standards compliance?
> 
> I can add a linter Github Action using Github's Super Linter to help for reporting linter errors but likely configuration files will be desired to suppress certain warnings which maintainers of this project will need to add if desired. If there's interest in this, I'll work on a PR for it.

I will leave this up to @igo95862 , but I can say with confidence that his approach is going to be merged one way or the other very soon.

**Comment by igo95862** [comment]
> Can you add a pyproject.toml for use with Poetry in the python root directory? Dependency management is far better with Poetry than pip (even the new pipx) https://github.com/python-poetry/poetry

There is are dependencies outside the standard library. I will look in to it if it can build packages better.

> Also were these run through pylint, mypy and black to ensure standards compliance?

I use mypy on `--strict` mode. Flake8 for formatting. Pylint has too many false positives in my experience compared to mypy.


BTW I am currently pretty busy with some other stuff. I am not sure I can put any updates any time soon.

**Comment by dankamongmen** [comment]
I'm thinking I'm gonna go ahead and merge this, but as /pythonc as opposed to /python, which we will retain until I can get everything equalized. Thanks so much for your contribution, @igo95862 ! I'll do this in the next few hours.

**Review by dankamongmen — APPROVED** [review]

**Comment by dankamongmen** [comment]
OK, I've moved `python` in my master to `cffi`, which ought clear the way for your work in `python`.

**Comment by dankamongmen** [comment]
ok, i'm going to go ahead and fork from your repo, and then create a PR with that, so i can clean up the conflicts and merge this.

**Comment by igo95862** [comment]
Sorry, what do I need to do?

**Comment by dankamongmen** [comment]
I've made https://github.com/dankamongmen/notcurses/tree/dankamongmen/python-igo, and simply moved over your `python/`, and committed it as one commit using `git commit --author`:

```
[schwarzgerat](0) $ git branch -l
  dankamongmen/clock_nanosleep_portable
  dankamongmen/pixel
* dankamongmen/python-igo
  dankamongmen/versiongrokkage
  dankamongmen/xcb
  igo95862-master
  master
[schwarzgerat](0) $ git log
commit 0d7e17f84c5efe952c5dcec18078413469913235 (HEAD -> dankamongmen/python-igo, origin/dankamongmen/python-igo)
Author: igo95862 <igo95862@yandex.ru>
Date:   Sun Nov 29 12:12:40 2020 -0500

    igo95862's new python bindings

```

**Comment by dankamongmen** [comment]
Gonna close this and create a new one with that branch, since this seems unmergeable. See #1154 .

**Pull request closed without merging**
